[config] Add support for accepting port alias in lieu of port name

paavaanan opened the pull request:
[portalias-logs.txt](https://github.com/Azure/sonic-utilities/files/2053930/portalias-logs.txt)
<!--
Please make sure you've read and understood our contributing guidelines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "closes #xxxx",
"fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
issue when the PR is merged

Please provide the following information:
-->

**- What I did**
Implemented portalias feature for two basic commands "config interface shutdown/startup"

-Introduced new CLI to configure the interface mode: 
  -config interface_mode "alias" 
  -config interface_mode "defaut"


User can stick to any mode. Mode selection is unique for each sessions. Default mode stick to sonic ethernet interface style and Alias mode stick to individual switch vendor interface style. User selection  sticks to the current logged session. If multiple login session do exists it maintains the user selection seperately.  

**- How I did it**

1. Introduced convert_interface_name() function to toggle interfaces between alias/default mode.
2. Included a termporary file for each individual session to maintain the mode selection.
3. interface_mode.sh script is included in /etc/profile.d to create a unique file for each session at the time of login.

**- How to verify it**
1. config interface_mode  alias 
2. config interface shutdown fortyGigE1/1/7
3. config interface_mode  default 
4. config interface startup Ethernet6    

--> interface_mode.sh is raised in https://github.com/Azure/sonic-buildimage/pull/1751

jleveque commented on config/main.py:
```diff
@@ -33,6 +35,64 @@ def run_command(command, display_cmd=False, ignore_error=False):
     if proc.returncode != 0 and not ignore_error:
         sys.exit(proc.returncode)
 
+def convert_interface_name(interface_name):
```
This function will blindly convert a port name to a port alias and vice-versa. This could pose issues if a device shares the same string for both a port name and a port alias, as it will always find a match on line 49 and return that port's alias.

I think it is better to have two functions with distinct functionality: `interface_name_to_alias(interface_name)` and `interface_alias_to_name(interface_alias)`.

paavaanan replied:
Introduced interface_alias_to_name() to convert the alias to default mode. In the current code change, there is no any need for interface_name_to_alias() conversion. While implementing the feature for other subset of commands if needed will implement on need basis.

paavaanan commented on config/main.py:
```diff
@@ -33,6 +35,64 @@ def run_command(command, display_cmd=False, ignore_error=False):
+
+def set_interface_mode(mode):
+    tmp = "/tmp/"
+    output = subprocess.check_output(['tty']).strip('\n')
```
Agreed. Initially we approched the issue in the above context. Since python is the child process of parent bash (as like for any sessios) we can inherit the status of enviroment variable say SONIC_CLI_ALIAS_MODE to python but we can't revert back the state to TRUE/FALSE back to bash environment variable. .  The change do exists until the script terminates. SONIC_CLI_ALIAS_MODE value cannot be exported to parent bash.

jleveque replied:
Ah, yes ... you can't modify the environment of the parent shell. That's a problem. If we wanted to proceed with this approach, I guess we would need to prompt the user to log out and log back in again for the change to take effect; I think that's quite an inconvenience.

@lguohan: Do you have any alternative suggestions? Otherwise, I believe the tty-based approach may be our best bet.

paavaanan commented on config/main.py:
```diff
@@ -33,6 +35,64 @@ def run_command(command, display_cmd=False, ignore_error=False):
     if proc.returncode != 0 and not ignore_error:
         sys.exit(proc.returncode)
 
+def interface_alias_to_name(interface_name):
```
Okay

paavaanan commented:
•	Modified the design to user based approach.   Harnessed the power of /etc/sudoers and /etc/environment file to retain the environment variable value even in sudo sessions. So, when the user request the change from default mode to alias mode, the action is completed in two phases. 
•	config/main.py script fetch the current value of IFMODE and modify as user requested alias mode and write in bashrc.
•	When the user logout and login the session bashrc update the environment variable.
•	This works for both root, and non-root sessions.

paavaanan commented:
- In this latest commit, addressed review comments.
- Maintained uniformity in naming convention.
- Environement variable IFMODE  alias/default switch code is refined.

[port_alias_config.log](https://github.com/Azure/sonic-utilities/files/2174959/port_alias_config.log)

jleveque commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+    set_mode = "IFMODE={}".format(mode)
+
+    if not sudo_user:
+        bashrc = "/{}/.bashrc".format(user)
```
The .bashrc path should always be `"/home/{}/.bashrc"`, whether or not the user has called the command with sudo.

I also believe you can simplify lines 75-82 as follows:

```
bashrc_ifmode_line = "IFMODE={}".format(mode)

user = os.getenv('SUDO_USER')

if not user:
    user = os.getenv('USER')

bashrc = "/home/{}/.bashrc".format(user)
...
```

paavaanan replied:
- What if root user try to do the change the .bashrc will be in /root/.bashrc. 
- Do we need to adress this scenario?

jleveque replied:
We never want to modify the .bashrc of the root user. This is the entire point of first checking to see if `SUDO_USER` is not empty. If SUDO_USER is not empty, we want to modify *that* user's .bashrc. Similarly, we want to check the environment variables of *that* user to determine the mode to use.

jleveque commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+    """
+    sudo_user = os.getenv('SUDO_USER')
+    user = os.getenv('USER')
+    set_mode = "IFMODE={}".format(mode)
```
- The environment variable should have a more descriptive name, like `SONIC_CLI_IFACE_MODE`

- `set_mode` should also have a more descriptive name, like `bashrc_ifacemode_line`.

- Also, you should append a newline at the end to ensure the file ends with a newline.

jleveque commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+    f.close()
+
+    if "IFMODE" not in filedata:
+        newdata = filedata + set_mode
```
You should probably prepend a newline here to be safe. If the file happens to end with no newline, the `IFMODE` line will be concatenated onto the existing final line.

jleveque commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+    if "IFMODE" not in filedata:
+        newdata = filedata + set_mode
+    else:
+        newdata = re.sub(r"IFMODE=\w+",set_mode,filedata)
```
Please add spaces after commas.

jleveque commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+    else:
+        newdata = re.sub(r"IFMODE=\w+",set_mode,filedata)
+
+    f = open(bashrc,'w')
```
Please add space after comma.

jleveque commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+    print "Please logout and login again!"
+    
+def get_interface_mode():
+    mode = os.getenv('IFMODE')
```
What if the environment variable is not set? You should probably return "default" in this case so that this function consistently returns a value.

paavaanan replied:
Will address this

jleveque commented on config/main.py:
```diff
@@ -556,5 +647,26 @@ def platform():
+
+@cli.group()
+def interface_mode():
+    """Interface mode change tasks"""
```
Suggest `"""Modify interface naming mode for interacting with SONiC CLI"""`

jleveque commented on config/main.py:
```diff
@@ -556,5 +647,26 @@ def platform():
+
+@interface_mode.command('default')
+def interface_mode_default():
+    """Set interface mode to DEFAULT"""
```
Suggest `"""Set CLI interface naming mode to DEFAULT (SONiC port name)"""`

jleveque commented on config/main.py:
```diff
@@ -556,5 +647,26 @@ def platform():
+
+@interface_mode.command('alias')
+def interface_mode_alias():
+    """Set interface mode to ALIAS"""
```
Suggest `"""Set CLI interface naming mode to ALIAS (Vendor port alias)"""`

jleveque commented on config/main.py:
```diff
@@ -556,5 +647,26 @@ def platform():
+#
+
+@cli.group()
+def interface_mode():
```
Suggest change to `interface_naming_mode()`

paavaanan commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+    f = open(bashrc,'w')
+    f.write(newdata)
+    f.close()
+    print "Please logout and login again!"
```
Agreed

paavaanan commented on config/main.py:
```diff
@@ -534,9 +532,9 @@ def interface():
-        interface_name = interface_alias_to_name(interface_name)
-    """Shut down interface"""
+    if get_interface_mode() == "alias": 
+        interface_name = interface_alias(interface_name)
```
If interface_alias() returns None. It tries to execute the "ip link <interface_name> down". This is totally invalid.
Why do this needs to be executed with in-appropriate name.

paavaanan commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+            if interface_name == port_name:
+                return port_dict[port_name]['alias']
+        print "Invalid interface {}".format(interface_name)
+        sys.exit(0)
```
```
def shutdown(interface_name, verbose):
    """Shut down interface"""
    print "lias"
    if get_interface_mode() == "alias":
        interface_name = interface_alias(interface_name)
```
- If we fail to return sys.exit(0) in function interface_alias() it propogates to next line and execute 
"ip link set <interface_nam> down. Where interface_name is invalid.
- This is unnecessary and tends to double opertaion. To avoid this sys.exit(0) introduced. and also it alerts the user about invalid interface.
- If you are skeptical to the exit in  utiltiy function we can shift the exit in parent shutdown().

```
    command = "ip link set {} down".format(interface_name)
    run_command(command, display_cmd=verbose)

```

#

jleveque replied:
Yes. I think the caller functions should check for `None`, print an error and exit.

paavaanan replied:
Will addres the changes

paavaanan commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
     if proc.returncode != 0 and not ignore_error:
         sys.exit(proc.returncode)
 
+def interface_alias(interface_name):
```
Thanks for the clarification

paavaanan commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+            if interface_name == port_dict[port_name]['alias']:
+                return port_name
+        print "Invalid interface {}".format(interface_name)
+        sys.exit(0)
```
comment addressed below

paavaanan commented on config/main.py:
```diff
@@ -33,6 +35,70 @@ def run_command(command, display_cmd=False, ignore_error=False):
+
+    port_dict = json.loads(p.stdout.read())
+
+    if interface_name is not None:
```
- The interface_name validation is done by pthon click package at the initial phase of the function it throws missing argument error.

```
admin@sonic:~$ sudo config interface shutdown None
Invalid interface None
````

```
admin@sonic:~$ sudo config interface shutdown
Usage: config interface shutdown [OPTIONS] <interface_name>
Error: Missing argument "interface_name".
```

jleveque replied:
I wasn't asking what happens if a user provides the string "None" as the interface name.

My question is regarding the fact that you have an `if` statement (`if interface_name is not None:`) which returns a value but no corresponding `else` statement. Thus, if `interface_name` is of type None (not the string "None"), this function will not return anything. I think this was overlooked, so I was asking for an explanation to see if this is the behavior you expect.

jleveque replied:
I wasn't asking what happens if a user provides the string "None" as the interface name.

My question is regarding the fact that you have an `if` statement (`if interface_name is not None:`) which returns a value but no corresponding `else` statement. Thus, if `interface_name` is of type None (not the string "None"), this function will not return anything. I think this was overlooked, so I was asking for an explanation to see if this is the behavior you expect.

paavaanan replied:
Joe I got your point.  When the interface_name variable type is None, even though else is not defined inside the corresponding function the return type of interface_alias_to_name() will be None.  All parent functions  are equipped to handle this case. So, Invalid Interface message will be raised to the user. Even if we add an else the behavior will be the same.

```def shutdown(interface_name, verbose):
    """Shut down interface"""
    if get_interface_mode() == "alias":
        interface_name = interface_alias_to_name(interface_name)
        if interface_name is None:
            print "Invalid interface {}".format(interface_name)<--------
            raise click.Abort()
)
```

```
admin@sonic:~$ sudo config interface shutdown hundredGigE1/1
Invalid interface None
Aborted!
admin@sonic:~$
```

jleveque replied:
I still believe that every path should explicitly return a value. Otherwise the behavior of the function isn't easily determined to the reader. I think you can easily fix this as follows:

```
    if interface_alias is not None:
        for port_name in natsorted(port_dict.keys()):
            if interface_alias == port_dict[port_name]['alias']:
                return port_name
        print "Invalid interface {}".format(interface_alias)
       
    return None
```

jleveque commented on config/main.py:
```diff
@@ -64,39 +63,42 @@ def interface_name_to_alias(interface_name):
-    if not sudo_user:
-        bashrc = "/{}/.bashrc".format(user)
+    user = os.getenv('SUDO_USER')
+    bashrc_ifacemode_line = "SONIC_CLI_IFACE_MODE={}".format(mode)
```
I think this line is best moved down just above line 87, right before it is used. That will also keep all of the user name determination code together.

jleveque commented on config/main.py:
```diff
@@ -64,39 +63,42 @@ def interface_name_to_alias(interface_name):
-        newdata = re.sub(r"IFMODE=\w+",set_mode,filedata)
-
+        newdata = re.sub(r"SONIC_CLI_IFACE_MODE=\w+", bashrc_ifacemode_line, filedata)
+    
```
Delete whitespace.

jleveque commented on config/main.py:
```diff
@@ -406,7 +408,12 @@ def add_vlan_member(ctx, vid, interface_name, untagged):
     
     if get_interface_mode() == "alias": 
-        interface_name = interface_alias(interface_name)
+        default_name = interface_alias_to_name(interface_name)
```
The introduction of the variable `default_name` is unnecessary. Simply assign directly to `interface_name` and compare it to None. You can then also eliminate the `else` clause:

```
interface_name = interface_alias_to_name(interface_name)
if interface_name is None:
    print "Invalid interface {}".format(interface_name)
    raise click.Abort()
```

jleveque commented on config/main.py:
```diff
@@ -406,7 +408,12 @@ def add_vlan_member(ctx, vid, interface_name, untagged):
-        interface_name = interface_alias(interface_name)
+        default_name = interface_alias_to_name(interface_name)
+        if default_name is None:
+            print "Invalid interface {}".format(interface_name)
```
`interface_name` is not defined. This will crash. However, instead of replacing it with `default_name`, please refactor the code as I suggested in my comment on line 411.

paavaanan replied:
How do this crash? interface_name is the incoming argument for add_vlan_member() So, argument do have a value when invoked. 
```
def add_vlan_member(ctx, vid, interface_name, untagged):
```

jleveque replied:
You're correct. Technically, Click should prevent `interface_name` from being `None`. I think I got a bit lost due the introduction of `default_interface`. Ignore this comment. #Closed

jleveque commented on config/main.py:
```diff
@@ -415,6 +422,11 @@ def add_vlan_member(ctx, vid, interface_name, untagged):
             alias_name = interface_name_to_alias(interface_name)
+            if alias_name is None:
+                print "Invalid interface {}".format(interface_name)
+                sys.exit(0)
```
Please replace `sys.exit(0)` `raise click.Abort()`, which will cause the program to exit with code 1.

jleveque commented on config/main.py:
```diff
@@ -435,15 +447,25 @@ def del_vlan_member(ctx, vid, interface_name):
+            alias_name = interface_name_to_alias(interface_name)
+            if alias_name is None:
+                print "Invalid interface {}".format(interface_name)
+                sys.exit(0)
```
Please replace `sys.exit(0)` `raise click.Abort()`, which will cause the program to exit with code 1.

jleveque commented on config/main.py:
```diff
@@ -435,15 +447,25 @@ def del_vlan_member(ctx, vid, interface_name):
 
     if get_interface_mode() == "alias": 
-        interface_name = interface_alias(interface_name)
+        default_name = interface_alias_to_name(interface_name)
```
Please refactor this code in the same manner as I suggested in my comment on line 411.

jleveque commented on config/main.py:
```diff
@@ -533,7 +555,12 @@ def interface():
     """Shut down interface"""
     if get_interface_mode() == "alias": 
-        interface_name = interface_alias(interface_name)
+        default_name = interface_alias_to_name(interface_name)
```
Please refactor this code in the same manner as I suggested in my comment on line 411.

jleveque commented on config/main.py:
```diff
@@ -548,7 +575,12 @@ def shutdown(interface_name, verbose):
     """Start up interface"""
     if get_interface_mode() == "alias": 
-        interface_name = interface_alias(interface_name)
+        default_name = interface_alias_to_name(interface_name)
```
Please refactor this code in the same manner as I suggested in my comment on line 411.

jleveque commented on config/main.py:
```diff
@@ -564,7 +596,12 @@ def startup(interface_name, verbose):
     """Set interface speed"""
     if get_interface_mode() == "alias": 
-        interface_name = interface_alias(interface_name)
+        default_name = interface_alias_to_name(interface_name)
```
Please refactor this code in the same manner as I suggested in my comment on line 411.

jleveque commented on portalias/PORTALIAS.md:
```diff
@@ -0,0 +1,75 @@
+#Portalias
```
This file is now very out-of-date, since we are no longer using a flag to determine the mode. Also, a design doc of this sort doesn't belong in this repo alongside the code; if anything it should be placed in the GitHub wiki. Please delete this file.

paavaanan commented:
@jleveque addressed all review comments and raised 2 commits with all changes. 
Please, have a look on this.

jleveque commented on config/main.py:
```diff
@@ -83,7 +83,7 @@ def set_interface_mode(mode):
         bashrc = "/home/{}/.bashrc".format(user)
     else:
-        sys.exit(0)
+        raise click.Abort
```
Add parentheses: `raise click.Abort()`

jleveque commented on config/main.py:
```diff
@@ -418,7 +418,7 @@ def add_vlan_member(ctx, vid, interface_name, untagged):
         interface_name = interface_alias_to_name(interface_name)
         if interface_name is None:
-            sys.exit(0)
+            raise click.Abort
```
Add parentheses: `raise click.Abort()`

jleveque commented on config/main.py:
```diff
@@ -453,7 +453,7 @@ def del_vlan_member(ctx, vid, interface_name):
         interface_name = interface_alias_to_name(interface_name)
         if interface_name is None:
-            sys.exit(0)
+            raise click.Abort
```
Add parentheses: `raise click.Abort()`

jleveque commented on config/main.py:
```diff
@@ -555,7 +555,7 @@ def shutdown(interface_name, verbose):
         interface_name = interface_alias_to_name(interface_name)
         if interface_name is None:
-            sys.exit(0)
+            raise click.Abort
```
Add parentheses: `raise click.Abort()`

jleveque commented on config/main.py:
```diff
@@ -572,7 +572,7 @@ def startup(interface_name, verbose):
         interface_name = interface_alias_to_name(interface_name)
         if interface_name is None:
-            sys.exit(0)
+            raise click.Abort
```
Add parentheses: `raise click.Abort()`

jleveque commented on config/main.py:
```diff
@@ -591,7 +591,7 @@ def speed(interface_name, interface_speed, verbose):
         interface_name = interface_alias_to_name(interface_name)
         if interface_name is None:
-            sys.exit(0)
+            raise click.Abort
```
Add parentheses: `raise click.Abort()`

jleveque commented on config/main.py:
```diff
@@ -33,6 +36,77 @@ def run_command(command, display_cmd=False, ignore_error=False):
+
+    if "SONIC_CLI_IFACE_MODE" not in filedata:
+        newdata = filedata + bashrc_ifacemode_line
+        newdata += "\n"
```
I suggest adding the newline directly to `bashrc_ifacemode_line` on line 78 as follows:

`bashrc_ifacemode_line = "SONIC_CLI_IFACE_MODE={}\n".format(mode)`
